enhancement of error message for output schema validation failure

mahesh-orch opened the pull request:
Output schema fails in validation if metayaml file has a parameter of a specific type and python script doesn't have equivalent type of it for same parameter. For an example:
In metayaml file:
```
output_schema:
  osTestParam:
    description: test parameter
    type: string
    required: true
    secret: true
```
In python script:
```
class SampleAction(Action):
    def run(self):
        osTestParam = None
        return {"osParam": osTestParam}
```
If this action ran from st2 web UI then it gives `// Action produced no data` output in action output box and nothing else rather than this. It is needful to have an appropriate error message on web UI when such case occurs.

m4dcoder commented on st2common/st2common/util/output_schema.py:
```diff
@@ -77,6 +79,61 @@ def mask_secret_output(ac_ex, output_value):
+                result["stderr"] = msg
+                result["stdout"] = ""
+                return (result, status)
+
```
This doesn't look right. How come the schema validation isn't able to pick up the type error and we have to manually check this with a bunch of if/elif statements? Can you research and provide a reasonable explanation here?

mahesh-orch replied:
<img width="815" alt="2021-08-05 (8)" src="https://user-images.githubusercontent.com/81608498/128306741-c6f1d640-8f68-4e01-99a7-88e6dc05f487.png">
<img width="832" alt="2021-08-05 (9)" src="https://user-images.githubusercontent.com/81608498/128306769-4775fad5-e9de-477a-9f3d-e1d908c28c48.png">
<img width="960" alt="2021-08-05 (4)" src="https://user-images.githubusercontent.com/81608498/128306785-f3b0f4d5-7428-4938-804a-165aeb9c40c7.png">

mahesh-orch replied:
<img width="960" alt="2021-08-05 (12)" src="https://user-images.githubusercontent.com/81608498/128311602-b358b880-6648-4d33-afcf-2da6c8f9dc4f.png">
<img width="960" alt="2021-08-05 (11)" src="https://user-images.githubusercontent.com/81608498/128311610-b35b2f5d-fd73-4d4b-9ae9-9e2f314d0d4e.png">
<img width="960" alt="2021-08-05 (10)" src="https://user-images.githubusercontent.com/81608498/128311618-7f05b225-1689-4cc9-970c-a686dbcdef08.png">
<img width="960" alt="2021-08-05 (13)" src="https://user-images.githubusercontent.com/81608498/128311624-8b3cd6aa-0758-4ac4-88fe-9997abc3a6e5.png">

mahesh-orch replied:
@m4dcoder As you said schema validation gives type error and gives message why output schema validation failed at CLI but when we run same action from web UI, it only gives `// Action produced no data` as output. As there is no error message, it isn't clear why action failed. After adding conditions for checking output schema property types in metadata and entry point files we can prepare appropriate result if any condition fails. It is giving proper error message at CLI and web UI as well.

mahesh-orch replied:
I have observed that these changes are causing key error in st2 pack install workflow. Please give me inputs for better approach to resolve this issue.

m4dcoder replied:
The if/else here is just wrong and heading in the wrong direction. The schema validation is doing the job of detecting the type error.  I would check the result with the error and message. Each runners have different format/structure for their result. I think the UI is expecting the error to be in a specific path in the result. Can you check this with the UI developer?  To fix the issue, you may have to make changes at https://github.com/StackStorm/st2/blob/master/st2common/st2common/util/output_schema.py#L95 to write the result correctly for the runner of the action.

m4dcoder replied:
The if/else here is just wrong and heading in the wrong direction. The schema validation is doing the job of detecting the type error. I would check the result with the error and message. Each runners have different format/structure for their result. I think the UI is expecting the error to be in a specific path in the result. Can you check this with the UI developer? To fix the issue, you may have to make changes at https://github.com/StackStorm/st2/blob/master/st2common/st2common/util/output_schema.py#L95 to write the result correctly for the runner of the action.

mahesh-orch commented:
@m4dcoder I have removed the several if/else statements. Have updated the result at https://github.com/StackStorm/st2/blob/master/st2common/st2common/util/output_schema.py#L95. It is giving proper error message in web UI as well. Also, added unit tests for changes. Had to make minor changes in existing unit tests in https://github.com/StackStorm/st2/blob/master/st2common/tests/unit/test_util_output_schema.py and https://github.com/StackStorm/st2/blob/master/contrib/runners/orquesta_runner/tests/unit/test_output_schema.py. Please review the changes. Thanks.

cognifloyd commented on st2common/st2common/util/output_schema.py:
```diff
@@ -93,8 +94,10 @@ def validate_output(runner_schema, action_schema, result, status, output_key):
         result = {
-            "error": str(ex),
-            "message": "Error validating output. See error output for more details.",
+            "exit_code": FAILURE_EXIT_CODE,
+            "result": "None",
+            "stderr": str(ex),
+            "stdout": "",
```
validation is not a subprocess, so using `exit_code`, `stderr`, and `stdout` feels a little odd here.
Can we adjust something in st2web to pick up the error message instead?

cognifloyd reviewed (changes requested):
I really like adding tests for output_schema validation!

I'm not sure about forcing this into a exit_code/stderr though. That is semantically wrong, esp from an orquesta workflow. If we need to tweak the output here (maybe call it `validation_error`?), I think that's cool. But, I think st2web is going to need massaging to recognize the validation error messages.

m4dcoder commented on contrib/runners/orquesta_runner/tests/unit/test_output_schema.py:
```diff
@@ -197,9 +197,12 @@ def test_fail_incorrect_output_schema(self):
+            "stdout": "",
+            "stderr": "Additional properties are not allowed",
+            "exit_code": 1,
+            "result": "None",
```
The structure of the expected result is different for each runner. What you have patched here is only valid for shell runner. I doubt this error will be caught for workflow execution. 1) Can you review each action runner (i.e. local shell, python, orquesta, etc.)? 2) Can you write a unit test for each action runner?

mahesh-orch replied:
@m4dcoder As per recent research it is looking like this problem can be solved with minor changes in UI. Let me try with changes in UI if they don't work I will surely work on reviewing and unit tests on each runner.

mahesh-orch commented:
@cognifloyd , @m4dcoder Done research with help of UI developer and found that for showing action output in web UI the currently we are picking up `result.result`, `result.stdout`, `result.stderr` and `result.traceback` only. As you both suggested, we will do changes in UI to pick up `result.error`.
